File: 123. Best Time to Buy and Sell Stock III/Python/Solution.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if not prices:
            return 0
        dp = [[0] * len(prices) for i in range(3)]
        for time in range(1, 3):
            current = dp[time - 1][0] - prices[0]
            for i in range(1, len(prices)):
                dp[time][i] = max(dp[time][i - 1], prices[i] + current)
                current = max(current, dp[time - 1][i] - prices[i])
        return max(dp[-1])

File: 123. Best Time to Buy and Sell Stock III/Java/Solution.java
public class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length == 0) {
            return 0;
        }
        int[][] dp = new int[3][prices.length];
        for (int time = 1; time < 3; time++) {
            int current = dp[time - 1][0] - prices[0];
            for (int i = 1; i < prices.length; i++) {
                dp[time][i] = Math.max(dp[time][i - 1], prices[i] + current);
                current = Math.max(current, dp[time - 1][i] - prices[i]);
            }
        }
        int[] last = dp[dp.length - 1];
        int max = last[0];
        for (int num : last) {
            if (max < num) {
                max = num;
            }
        }
        return max;
    }
}